File: scripts/reel-frame.py
#!/usr/bin/env python3
"""Render the reel's 1080x1920 frames: the intro card, and the per-clip overlay
that carries the keyword header and the burned-in subtitle.

This exists because the ffmpeg on this machine is built without libfreetype and
without libass — there is no `drawtext` and no `subtitles` filter, so no text can
be drawn inside the filter graph. Pillow draws it instead.

That turns out to be the better path anyway: Pillow reports exact glyph metrics,
so each furigana cluster is positioned by measuring the kanji run it sits above.
A space-padded ruby line drifts in any proportional font; this does not.

The overlay is RGBA with a fully transparent rectangle punched where the clip
plays, so ffmpeg composites video underneath in one pass. The pane rect is
printed as JSON for the merge script to scale the clip into.

  reel-frame.py '<spec json>'      spec.mode = "card" | "overlay"
"""
import json
import sys
from PIL import Image, ImageDraw, ImageFont

W, H = 1080, 1920

BG = (7, 10, 15, 255)
MINT = (200, 230, 201, 255)
RED = (230, 57, 70, 255)
INK = (29, 53, 87, 255)
MID = (55, 65, 81, 255)
SOFT = (91, 107, 122, 255)
WHITE = (244, 247, 250, 255)
GREY = (150, 160, 174, 255)
RULE = (29, 53, 87, 70)

# Header band, then the video pane, then the subtitle block.
BAND = (40, 52, 1040, 428)          # x0, y0, x1, y1
PANE = (0, 470, 1080, 1078)         # 1080x608 — 16:9 at full width
TEXT_TOP = 1130

JP = "/System/Library/Fonts/Hiragino Sans GB.ttc"
SERIF = "/System/Library/Fonts/NewYork.ttf"
SANS = "/System/Library/Fonts/HelveticaNeue.ttc"
BN = "/System/Library/Fonts/KohinoorBangla.ttc"


def font(path, size, index=0):
    try:
        return ImageFont.truetype(path, size, index=index)
    except Exception:
        return ImageFont.load_default()


def width(d, text, f):
    if not text:
        return 0
    box = d.textbbox((0, 0), text, font=f)
    return box[2] - box[0]


def centre(d, y, text, f, fill, cx=W / 2):
    """Draw `text` horizontally centred on `cx`; returns the y below it."""
    if not text:
        return y
    box = d.textbbox((0, 0), text, font=f)
    d.text((cx - (box[2] - box[0]) / 2 - box[0], y - box[1]), text, font=f, fill=fill)
    return y + (box[3] - box[1])


def wrap(d, text, f, maxw):
    words, lines, cur = text.split(), [], ""
    for w_ in words:
        trial = f"{cur} {w_}".strip()
        if width(d, trial, f) <= maxw or not cur:
            cur = trial
        else:
            lines.append(cur)
            cur = w_
    if cur:
        lines.append(cur)
    return lines


# ── furigana ────────────────────────────────────────────────────────────────

def is_kanji(c):
    return '一' <= c <= '鿿' or '㐀' <= c <= '䶿' or c in '々〆〇'


def parse_jp(s):
    """`漢字(かんじ)` markup → [(surface, reading|'')], matching the Studio's parser."""
    out, buf = [], ""
    i = 0
    while i < len(s):
        c = s[i]
        if c in '(（':
            j, rd = i + 1, ""
            while j < len(s) and s[j] not in ')）':
                rd += s[j]
                j += 1
            k = len(buf)
            while k > 0 and is_kanji(buf[k - 1]):
                k -= 1
            base, prefix = buf[k:], buf[:k]
            if not base:
                base, prefix = buf, ""
            if prefix:
                out.append((prefix, ""))
            buf = ""
            if base:
                out.append((base, rd.strip()))
            i = j + 1
        else:
            buf += c
            i += 1
    if buf:
        out.append((buf, ""))
    return [t for t in out if t[0]]


def draw_japanese(d, y, markup, keyword, size):
    """Ruby line: each reading is centred over the kanji run it belongs to."""
    toks = parse_jp(markup)
    f_base = font(JP, size, 1)
    f_ruby = font(JP, max(18, int(size * 0.40)), 0)

    widths = [width(d, s, f_base) for s, _ in toks]
    total = sum(widths)
    x = (W - total) / 2
    ruby_h = int(size * 0.40) + 10
    base_y = y + ruby_h

    for (surface, reading), w_ in zip(toks, widths):
        hit = bool(keyword) and (keyword in surface or surface in keyword and len(surface) > 1)
        d.text((x, base_y), surface, font=f_base, fill=RED if hit else WHITE)
        if reading:
            rw = width(d, reading, f_ruby)
            d.text((x + (w_ - rw) / 2, y), reading, font=f_ruby, fill=RED if hit else GREY)
        x += w_
    return base_y + int(size * 1.25)


def fit_japanese(d, markup, start_size, maxw):
    """Shrink until the line fits the frame rather than letting it run off."""
    size = start_size
    while size > 30:
        f = font(JP, size, 1)
        if sum(width(d, s, f) for s, _ in parse_jp(markup)) <= maxw:
            return size
        size -= 4
    return size


def draw_romaji(d, y, romaji, key_romaji, f):
    """Same line, but the searched word's romaji is picked out in red."""
    parts = romaji.split()
    if not parts:
        return y
    space = width(d, ' ', f)
    total = sum(width(d, p, f) for p in parts) + space * (len(parts) - 1)
    x = (W - total) / 2
    top = None
    for p in parts:
        box = d.textbbox((0, 0), p, font=f)
        if top is None:
            top = box[1]
        d.text((x, y - box[1]), p, font=f, fill=RED if p == key_romaji else GREY)
        x += width(d, p, f) + space
    return y + (d.textbbox((0, 0), 'Ay', font=f)[3])


# ── the header card ─────────────────────────────────────────────────────────

def draw_header(d, box, spec, big):
    """The mint keyword card. `big` renders it full-frame for the intro."""
    x0, y0, x1, y1 = box
    d.rounded_rectangle(box, radius=30, fill=MINT)

    romaji = (spec.get('romaji') or '').strip()
    meaning = (spec.get('meaningEn') or '').strip()
    reading = (spec.get('reading') or '').strip()
    word = (spec.get('word') or '').strip()

    scale = 1.0 if not big else 1.5
    f_badge = font(JP, int(38 * scale), 1)
    f_brand = font(SANS, int(21 * scale), 2)
    f_romaji = font(SERIF, int(74 * scale))
    f_mean = font(SERIF, int(34 * scale))
    f_read = font(JP, int(36 * scale), 0)
    word_size = int((96 if len(word) <= 4 else 76) * scale)
    f_word = font(JP, word_size, 1)

    # Brand mark, top-left of the card.
    bx, by = x0 + 38, y0 + 40
    d.rounded_rectangle([bx, by, bx + int(76 * scale), by + int(76 * scale)], radius=int(16 * scale), fill=(255, 255, 255, 255))
    centre(d, by + int(18 * scale), '文', f_badge, RED, cx=bx + int(38 * scale))
    centre(d, by + int(90 * scale), 'Learn Japanese', f_brand, INK, cx=bx + int(38 * scale))
    centre(d, by + int(116 * scale), 'with anime', f_brand, INK, cx=bx + int(38 * scale))

    cx = (x0 + x1) / 2 + int(60 * scale)
    inner = int(300 * scale)
    y = y0 + int(34 * scale)
    y = centre(d, y, romaji, f_romaji, INK, cx=cx) + int(14 * scale)
    d.line([cx - inner, y, cx + inner, y], fill=RULE, width=2)
    y += int(16 * scale)
    y = centre(d, y, meaning, f_mean, MID, cx=cx) + int(14 * scale)
    d.line([cx - inner, y, cx + inner, y], fill=RULE, width=2)
    y += int(16 * scale)
    y = centre(d, y, reading, f_read, SOFT, cx=cx) + int(20 * scale)
    centre(d, y, word, f_word, RED, cx=cx)


# ── modes ───────────────────────────────────────────────────────────────────

def render_card(spec, out):
    img = Image.new('RGBA', (W, H), MINT)
    d = ImageDraw.Draw(img)
    draw_header(d, (60, 520, W - 60, 1400), spec, big=True)
    img.convert('RGB').save(out)


def render_overlay(spec, out):
    img = Image.new('RGBA', (W, H), BG)
    d = ImageDraw.Draw(img)
    draw_header(d, BAND, spec, big=False)

    line = spec.get('line') or {}
    keyword = (spec.get('word') or '').strip()
    key_romaji = (spec.get('romaji') or '').strip()

    f_en = font(SANS, 34, 2)
    f_ro = font(SANS, 32, 0)
    f_bn = font(BN, 44, 2)

    y = TEXT_TOP
    for ln in wrap(d, (line.get('english') or '').strip(), f_en, W - 140)[:2]:
        y = centre(d, y, ln, f_en, GREY) + 12
    y += 26

    jp = (line.get('japanese_furigana') or '').strip()
    if jp:
        y = draw_japanese(d, y, jp, keyword, fit_japanese(d, jp, 64, W - 90))

    y = draw_romaji(d, y + 8, (line.get('romaji') or '').strip(), key_romaji, f_ro) + 34

    for ln in wrap(d, (line.get('bangla') or '').strip(), f_bn, W - 140)[:2]:
        y = centre(d, y, ln, f_bn, WHITE) + 16

    # Punch the video pane out last so nothing is drawn over the clip.
    d.rectangle(PANE, fill=(0, 0, 0, 0))
    img.save(out)


def main():
    spec = json.loads(sys.argv[1])
    out = spec['out']
    if spec.get('mode') == 'overlay':
        render_overlay(spec, out)
    else:
        render_card(spec, out)
    print(json.dumps({'out': out, 'pane': [PANE[0], PANE[1], PANE[2] - PANE[0], PANE[3] - PANE[1]]}))


if __name__ == '__main__':
    main()

File: scripts/reel-frame.mjs
/**
 * Renders the reel's 1080x1920 frames — the intro card and the per-clip overlay
 * carrying the keyword header and the burned-in subtitle.
 *
 * Rendered in headless Chrome rather than in ffmpeg or Pillow, for two reasons
 * that are both hard blockers rather than preferences:
 *
 *   1. This ffmpeg is built without libfreetype and without libass, so there is
 *      no `drawtext` and no `subtitles` filter — no text can be drawn in the
 *      filter graph at all.
 *   2. This Pillow reports `raqm: False`, meaning no HarfBuzz shaping. Bengali
 *      is a complex script: without shaping, কর + ো renders as a dotted-circle
 *      placeholder instead of করো. Japanese survives it; Bangla does not.
 *
 * Chrome shapes Bengali correctly and lays out `<ruby>` natively, so furigana
 * sits over its own kanji cluster with no width measuring at all.
 *
 * The overlay is a PNG with a transparent band where the clip plays, so ffmpeg
 * composites the video underneath in a single pass.
 */
import { spawn } from 'node:child_process'
import { writeFile, mkdtemp, rm, stat } from 'node:fs/promises'
import { existsSync } from 'node:fs'
import { tmpdir } from 'node:os'
import path from 'node:path'

export const FRAME = { w: 1080, h: 1920 }
/** One reel per language; each carries exactly one translation row. */
export const LANGS = ['en', 'bn', 'vi', 'ne']
export const LANG_NAMES = { en: 'English', bn: 'Bangla', vi: 'Vietnamese', ne: 'Nepali' }
/** Top block, video pane, bottom block — must sum to FRAME.h exactly. */
export const LAYOUT = { top: 470, pane: 608, bottom: 842 }
export const PANE_Y = LAYOUT.top

const CHROME_CANDIDATES = [
  '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome',
  '/Applications/Chromium.app/Contents/MacOS/Chromium',
  '/Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge',
  '/Applications/Brave Browser.app/Contents/MacOS/Brave Browser',
]

export function findChrome() {
  return CHROME_CANDIDATES.find(p => existsSync(p)) || null
}

const esc = (s) => String(s ?? '').replace(/[&<>"]/g, c => (
  { '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;' }[c]
))

const isKanji = (c) => /[一-鿿㐀-䶿々〆〇]/.test(c)

/** `漢字(かんじ)` markup → tokens, matching src/subtitles/timeline.ts's parser. */
function parseJP(str) {
  const out = []
  let buf = ''
  for (let i = 0; i < (str || '').length; i++) {
    const c = str[i]
    if (c === '(' || c === '（') {
      let j = i + 1, rd = ''
      while (j < str.length && str[j] !== ')' && str[j] !== '）') rd += str[j++]
      let k = buf.length
      while (k > 0 && isKanji(buf[k - 1])) k--
      let base = buf.slice(k)
      const prefix = buf.slice(0, k)
      if (!base) base = buf
      if (base !== buf && prefix) out.push({ s: prefix, f: '' })
      buf = ''
      if (base) out.push({ s: base, f: rd.trim() })
      i = j
    } else buf += c
  }
  if (buf) out.push({ s: buf, f: '' })
  return out.filter(t => t.s.length > 0)
}

/** Ruby markup, with the searched word picked out. */
function japaneseHtml(markup, keyword) {
  return parseJP(markup).map(t => {
    const hit = keyword && (t.s.includes(keyword) || (keyword.includes(t.s) && t.s.length > 1))
    const cls = hit ? ' class="key"' : ''
    const body = t.f
      ? `<ruby${cls}>${esc(t.s)}<rt>${esc(t.f)}</rt></ruby>`
      : `<span${cls}>${esc(t.s)}</span>`
    return body
  }).join('')
}

function romajiHtml(romaji, keyRomaji) {
  return (romaji || '').split(/\s+/).filter(Boolean)
    .map(p => (p === keyRomaji ? `<span class="key">${esc(p)}</span>` : esc(p)))
    .join(' ')
}

const CSS = `
*{margin:0;padding:0;box-sizing:border-box}
html,body{width:${FRAME.w}px;height:${FRAME.h}px;background:transparent}
body{
  font-family:'Helvetica Neue',Helvetica,Arial,sans-serif;
  -webkit-font-smoothing:antialiased;
  display:flex;flex-direction:column;
}
.jp{font-family:'Hiragino Sans','Hiragino Kaku Gothic ProN','Noto Sans JP',sans-serif}
.serif{font-family:'Times New Roman',Georgia,serif}
.key{color:#E63946}

/* One translation row per reel, in that language's own script. */
.t-en,.t-vi{font-family:'Helvetica Neue',Helvetica,Arial,sans-serif}
.t-bn{font-family:'Kohinoor Bangla','Bangla MN','Noto Sans Bengali',sans-serif}
.t-ne{font-family:'Kohinoor Devanagari','Devanagari Sangam MN','Noto Sans Devanagari',sans-serif}

.top{height:${LAYOUT.top}px;background:#070A0F;padding:52px 40px 0}
.pane{height:${LAYOUT.pane}px;background:transparent}
.bottom{
  height:${LAYOUT.bottom}px;background:#070A0F;
  padding:56px 56px 0;display:flex;flex-direction:column;
  align-items:center;justify-content:center;gap:26px;text-align:center;
}

.card{
  background:#C8E6C9;border-radius:30px;height:100%;
  display:flex;align-items:center;gap:28px;padding:26px 40px;
}
.brand{width:190px;flex:none;text-align:center;color:#1D3557}
.badge{
  width:78px;height:78px;margin:0 auto 12px;border-radius:18px;background:#fff;
  display:flex;align-items:center;justify-content:center;
  font-size:40px;font-weight:700;color:#E63946;
}
.brand p{font-size:21px;font-weight:700;line-height:1.3}
.stack{flex:1;text-align:center}
.stack .romaji{font-size:78px;line-height:1.05;color:#1D3557}
.stack .rule{height:1px;background:rgba(29,53,87,.32);margin:12px auto;width:78%}
.stack .meaning{font-size:35px;color:#374151}
.stack .kana{font-size:37px;color:#5B6B7A;margin-top:4px}
.stack .word{font-size:var(--word-size);font-weight:700;color:#E63946;line-height:1.15;margin-top:6px}

.line{font-size:var(--jp-size);font-weight:700;color:#F4F7FA;line-height:1.6;max-width:1000px}
.line rt{font-size:.34em;font-weight:500;color:#96A0AE;line-height:1}
.line .key rt{color:#E63946}
.ro{font-size:33px;color:#96A0AE;letter-spacing:.01em;max-width:960px}
.tr{font-size:var(--tr-size);font-weight:600;color:#F4F7FA;line-height:1.5;max-width:980px}

/* Intro card: the same card, given the whole frame. */
.solo{height:${FRAME.h}px;background:#C8E6C9;display:flex;align-items:center;padding:0 60px}
.solo .card{height:auto;padding:70px 56px;box-shadow:0 30px 70px rgba(15,23,41,.10)}
.solo .brand{width:250px}
.solo .badge{width:104px;height:104px;font-size:54px}
.solo .brand p{font-size:27px}
.solo .stack .romaji{font-size:112px}
.solo .stack .meaning{font-size:46px}
.solo .stack .kana{font-size:50px}
.solo .stack .word{font-size:calc(var(--word-size) * 1.5)}
`

function cardHtml(spec) {
  const word = String(spec.word || '')
  const size = word.length <= 2 ? 104 : word.length <= 4 ? 84 : 64
  return `<div class="card" style="--word-size:${size}px">
  <div class="brand">
    <div class="badge jp">文</div>
    <p>Learn Japanese<br>with anime</p>
  </div>
  <div class="stack">
    <div class="romaji serif">${esc(spec.romaji)}</div>
    <div class="rule"></div>
    <div class="meaning serif">${esc(spec.meaningEn)}</div>
    <div class="rule"></div>
    <div class="kana jp">${esc(spec.reading)}</div>
    <div class="word jp">${esc(word)}</div>
  </div>
</div>`
}

function page(body) {
  return `<!doctype html><html><head><meta charset="utf-8"><style>${CSS}</style></head><body>${body}</body></html>`
}

export function buildHtml(spec) {
  if (spec.mode === 'card') {
    return page(`<div class="solo">${cardHtml(spec)}</div>`)
  }
  const line = spec.line || {}
  const lang = LANGS.includes(spec.lang) ? spec.lang : 'en'
  const jp = String(line.japanese_furigana || '')
  // Long lines shrink rather than wrapping into a second ruby row.
  const plain = parseJP(jp).reduce((n, t) => n + t.s.length, 0)
  const jpSize = plain <= 16 ? 66 : plain <= 22 ? 58 : plain <= 28 ? 50 : 44
  // Bengali and Devanagari carry more vertical detail than Latin at the same
  // point size, and both wrap sooner, so they get their own scale.
  const translation = String(line.translation || '')
  const trBase = lang === 'bn' || lang === 'ne' ? 46 : 42
  const trSize = translation.length <= 46 ? trBase : translation.length <= 80 ? trBase - 6 : trBase - 11
  return page(`
<div class="top">${cardHtml(spec)}</div>
<div class="pane"></div>
<div class="bottom" style="--jp-size:${jpSize}px;--tr-size:${trSize}px">
  ${jp ? `<div class="line jp">${japaneseHtml(jp, spec.word)}</div>` : ''}
  ${line.romaji ? `<div class="ro">${romajiHtml(line.romaji, spec.romaji)}</div>` : ''}
  ${translation ? `<div class="tr t-${lang}">${esc(translation)}</div>` : ''}
</div>`)
}

/** Screenshot one frame. `mode: 'overlay'` keeps the video pane transparent. */
export async function renderFrame(spec, outPng, chrome) {
  const dir = await mkdtemp(path.join(tmpdir(), 'reel-frame-'))
  const html = path.join(dir, 'frame.html')
  await writeFile(html, buildHtml(spec), 'utf8')
  const args = [
    '--headless=new', '--disable-gpu', '--hide-scrollbars', '--no-sandbox',
    // A dedicated profile: without it Chrome tries to talk to an already-running
    // instance and the screenshot never returns.
    `--user-data-dir=${path.join(dir, 'profile')}`,
    '--no-first-run', '--no-default-browser-check', '--disable-extensions',
    '--disable-background-networking', '--disable-sync', '--disable-default-apps',
    '--force-device-scale-factor=1',
    '--virtual-time-budget=4000',
    `--window-size=${FRAME.w},${FRAME.h}`,
    `--screenshot=${outPng}`,
  ]
  if (spec.mode === 'overlay') args.push('--default-background-color=00000000')
  args.push(`file://${html}`)

  try {
    await new Promise((resolve, reject) => {
      const proc = spawn(chrome, args, { stdio: 'ignore' })
      let settled = false
      const finish = (err) => {
        if (settled) return
        settled = true
        clearInterval(poll)
        try { proc.kill('SIGKILL') } catch { /* already gone */ }
        err ? reject(err) : resolve()
      }
      // Chrome writes the screenshot but does not always exit afterwards, so
      // wait for the file to appear and settle rather than for the process.
      let lastSize = -1
      let waited = 0
      const poll = setInterval(async () => {
        waited += 200
        try {
          const { size } = await stat(outPng)
          if (size > 0 && size === lastSize) return finish()
          lastSize = size
        } catch { /* not written yet */ }
        if (waited > 30000) finish(new Error('Chrome timed out rendering a frame'))
      }, 200)
      proc.on('error', finish)
      proc.on('close', code => {
        if (settled) return
        if (code !== 0 && lastSize <= 0) finish(new Error(`Chrome exited ${code}`))
      })
    })
  } finally {
    await rm(dir, { recursive: true, force: true })
  }
  return outPng
}
